Serialization for per channel qtensor

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* **#26339 Serialization for per channel qtensor**

Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

Commit:
Serialization for per channel qtensor

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

[ghstack-poisoned]

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -403,8 +404,41 @@ void Pickler::pushLiteralTensor(const IValue& ivalue) {
-    pushInt(tensor.q_zero_point());
+    push<PickleOpCode>(PickleOpCode::MARK);
+    pushGlobal("torch", toString(tensor.qscheme()));
+    // tuple of (qscheme, scale, zp) or (qscheme, scales, zps, axis)
```
Why are there two different schemas here? Is it possible to always do it as `(qscheme, scales, zps, axis)`? Specializing is going to make the (de)serialization logic much more complicated

dzhulgakov replied:
It depends on quantizer (which can be per-tensor or per-channel) - the per-tensor one doesn't have axis. I can make axis to be None in the other case, but not sure it makes it cleaner. Specialization still has to be there as the scale vs scales is int vs tuple of int

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -403,8 +404,41 @@ void Pickler::pushLiteralTensor(const IValue& ivalue) {
+      case at::kPerChannelAffine: {
+        const auto* quantizer = static_cast<at::PerChannelAffineQuantizer*>(
+            tensor.quantizer().get());
+        push<PickleOpCode>(PickleOpCode::MARK);
```
You can shorten all of these by creating a `c10::tuple` and using `pushIValue`

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -812,18 +846,53 @@ PickleOpCode Unpickler::readInstruction() {
+                  axis.push_back(x.toInt());
+                }
+                result = _empty_per_channel_affine_quantized(
+                    {0}, scales, zero_points, axis, storage_tensor.options());
```
Why are all the parameters for `_empty_per_channel_affine_quantized` `Tensor`s instead of a list?

jerryzh168 replied:
because `scales` is float array, we don't have float array and we don't have plans to support it

dzhulgakov replied:
also, very theoretically, we might have multi-axis quantization in which case scales would be a multidimensional tensor. It's good to have a support for it in api, but it's not likely to appear soon. (note that in serialiation it's still fine to fold the scales to a tuple as the shape is deducible from the original tensor.

driazati reviewed (commented):
I'm generally apprehensive on the serialization changes, there's a lot of special casing / deeply paths in the tensor serialization and it's going to be hard to ensure there's no forward/backward compatibility problems.

driazati commented:
As for tests that compare, you can try to save with the pickler (e.g. use `torch.save` in a `@script` function) and load it in eager, and vice versa. This test does that for regular tensors:


https://github.com/pytorch/pytorch/blob/06c69ad8ede8aa90a035da6405a3267fa5efe915/test/test_jit.py#L14797

jerryzh168 commented on test/common_utils.py:
```diff
@@ -722,11 +722,21 @@ def assertTensorsEqual(a, b):
             elif x.is_quantized and y.is_quantized:
```
nit: we can remove `and y.is_quantized` here

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -208,13 +208,29 @@ def test_qtensor_per_channel_permute(self):
+        zero_points = torch.round(torch.rand(10) * 20 + 1).to(torch.long)
+        # quint32 is not supported yet
+        for dtype in [torch.quint8, torch.qint8]:
+            qr = torch.quantize_linear_per_channel(r, scales, zero_points, [1], dtype)
```
what is the behavior for slicing(view) on per channel quantized tensor right now?

dzhulgakov replied:
any slicing / set strides operation today basically errors out with nice message of "can't do it".  We could support a subset of these operations, but I doubt it's worth it at this point.

raghuramank100 replied:
Is this true for permute also? We could conceptually support it, but again doesn't  seem like it is really needed

dzhulgakov replied:
yes, `permute` is implemented as strides setting, so it gives the same error

jerryzh168 reviewed (approved):
checked code except `jit/pickler.cpp`.

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

[ghstack-poisoned]

raghuramank100 commented on torch/tensor.py:
```diff
@@ -41,14 +41,27 @@ def __reduce_ex__(self, proto):
         # See Note [Don't serialize hooks]
         torch.utils.hooks.warn_if_has_hooks(self)
         if self.is_quantized:
+            if self.qscheme() == torch.per_tensor_affine:
```
We still support torch.per_tensor_symmetric and torch.per_channel_symmetric as valid qschemes. They are used only by the observer currently. We should add checks for these too whenever we have a qscheme based check.

dzhulgakov replied:
this code will never be hit by observers - you can't create a tensor with that quantizer

raghuramank100 commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -786,18 +808,46 @@ PickleOpCode Unpickler::readInstruction() {
+                result = at::_empty_affine_quantized(
+                    {0}, storage_tensor.options(), q_scale, q_zero_point);
+              } break;
+              case at::kPerChannelAffine: {
```
We should also handle kPerChannelSymmetric and kPerTensorSymmetric here

dzhulgakov replied:
again, we have only two types of quantizers present today

raghuramank100 commented on test/test_jit.py:
```diff
@@ -3297,23 +3297,29 @@ def test_print_op_module(self):
-        buffer.seek(0)
-        model_loaded = torch.jit.load(buffer)
-        self.assertEqual(model_loaded(), model())
+        for per_channel in [False, True]:
```
Can we also sweep over qschemes in this test?

dzhulgakov replied:
@raghuramank100 - it's already sweeping among the two qschemes we support. We don't currently pass qscheme to any API and there are two calls quantize_linear and quantize_linear_per_channel to create tensor, so I don't think we can create symmetric tensor today

raghuramank100 commented on test/test_jit.py:
```diff
@@ -14841,6 +14847,34 @@ def forward(self, x):
+                    torch.save((x, y), self.fname)
+                    return y
+
+            q = torch.quantize_linear(
```
Suggest also sweeping over qschemes

dzhulgakov replied:
it does sweep on the two we support (See `qc` below)

raghuramank100 reviewed (changes requested):
Please add support for symmetric qschemes.

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]

Commit:
Update on "Serialization for per channel qtensor"


Serializes per-channel tensor in both torch.serialization and jit. Since we didn't bind Quantizer properly yet, I chose to save a tuple representing quantizer settings. To avoid recursive tensor serialization calls, I'm using tuple instead of tensor to store scales and zero points.

@driazati - please check the serialization logic. Is there a good test that compares that JIT serialization and python serialization are equivalent? (I haven't tested it yet)

Differential Revision: [D17443222](https://our.internmc.facebook.com/intern/diff/D17443222)

[ghstack-poisoned]
